morpho-checkpoint-node/src/Morpho/Config/Types.hs

```diff
@@ -38,133 +43,188 @@ import Morpho.Config.Orphans ()
+    ncNodeHost :: Wear w f NodeHostAddress,
+    ncNodePort :: Wear w f PortNumber,
+    ncValidateDb :: Wear w f Bool
   }
-  deriving (Show, Eq)
-
-instance FromJSON NodeConfiguration where
-  parseJSON = withObject "NodeConfiguration" $ \v -> do
-    nId <- v .: "NodeId"
-    ptcl <- v .: "Protocol"
-    numCoreNode <- v .: "NumCoreNodes"
-    rNetworkMagic <- v .: "RequiresNetworkMagic"
-    networkMagic <- v .: "NetworkMagic"
-    systemStart <- v .:? "SystemStart"
-    securityParam <- v .: "SecurityParam"
-    stableLedgerDepth <- v .: "StableLedgerDepth"
-    loggingSwitch <- v .: "TurnOnLogging"
-    traceOptions <- traceConfigParser v
-    vMode <- v .: "ViewMode"
-    logMetrics <- v .: "TurnOnLogMetrics"
-    -- Update Parameters
-    lkBlkVersionMajor <- v .: "LastKnownBlockVersion-Major"
-    lkBlkVersionMinor <- v .: "LastKnownBlockVersion-Minor"
-    lkBlkVersionAlt <- v .: "LastKnownBlockVersion-Alt"
-    slotLength <- v .: "SlotDuration"
-    snapshotsOnDisk <- v .: "SnapshotsOnDisk"
-    snapshotInterval <- v .: "SnapshotInterval"
-    blockFetchInterval <- v .:? "PoWBlockFetchInterval"
-    powNodeRpcUrl <- v .: "PoWNodeRpcUrl"
-    promPort <- v .: "PrometheusPort"
-    -- Checkpointing parameters
-    checkpointInterval <- v .: "CheckpointInterval"
-    requiredMajority <- v .: "RequiredMajority"
-    fedPubKeys <- v .: "FedPubKeys"
-    nodePrivKeyFile <- v .: "NodePrivKeyFile"
-    pure $
-      NodeConfiguration
-        ptcl
-        (CoreId (CoreNodeId nId))
-        numCoreNode
-        rNetworkMagic
-        networkMagic
-        systemStart
-        securityParam
-        stableLedgerDepth
-        loggingSwitch
-        traceOptions
-        logMetrics
-        vMode
-        ( Update
-            ( LastKnownBlockVersion
-                lkBlkVersionMajor
-                lkBlkVersionMinor
-                lkBlkVersionAlt
-            )
-        )
-        slotLength
-        snapshotsOnDisk
-        snapshotInterval
-        blockFetchInterval
-        powNodeRpcUrl
-        promPort
-        checkpointInterval
-        requiredMajority
-        fedPubKeys
-        nodePrivKeyFile
+  deriving (Generic)
+
+instance FunctorB (NodeConfiguration_ Covered)
+
+instance ApplicativeB (NodeConfiguration_ Covered)
+
+instance TraversableB (NodeConfiguration_ Covered)
+
+instance ConstraintsB (NodeConfiguration_ Covered)
+
+deriving instance AllBF Eq f (NodeConfiguration_ Covered) => Eq (NodeConfiguration_ Covered f)
+
+deriving instance AllBF Show f (NodeConfiguration_ Covered) => Show (NodeConfiguration_ Covered f)
+
+instance BareB NodeConfiguration_
+
+instance FunctorB (NodeConfiguration_ Bare)
+
+instance ConstraintsB (NodeConfiguration_ Bare)
+
+deriving instance AllBF Eq f (NodeConfiguration_ Bare) => Eq (NodeConfiguration_ Bare f)
+
+deriving instance AllBF Show f (NodeConfiguration_ Bare) => Show (NodeConfiguration_ Bare f)
+
+type NodeConfiguration = NodeConfiguration_ Bare Identity
```

Comment:
Actually that makes me think that the whole Bare/Covered thing is a bit overkill, because the same can be achieved with just an `HKD` type family as described in [this blog post](https://reasonablypolymorphic.com/blog/higher-kinded-data/):

```haskell
type family HKD f a where
  HKD Identity a = a
  HKD f        a = f a
```

Then the Covered/Bare argument can be completely removed from the type. I'll try that in a new commit

Reply:
Unfortunately this turns out to not work as easily as I'd hoped. Reason being that we can't just derive `FunctorB` instances anymore, since the `HKD` type family "hides" the underlying type of the fields, and `FunctorB` then doesn't know how to handle them. And afaik there's no way to tell GHC to make an instance for both `Identity` and "not"-`Identity`. I guess that's the whole reason why `Wear` exists, to make this easily possible.

So I think I'll use your suggestion instead, to have a type alias to hide the barbies part.